client/core/core.go

```diff
@@ -3820,6 +3820,19 @@ func (c *Core) resumeTrades(dc *dexConnection, trackers []*trackedTrade) assetMa
 				}
 				counterTxData := metaData.Proof.CounterTxData
 
+				// TODO: We will need careful handling here. To be clear, the
+				// client should reject orders at the funding stage when the
+				// server doesn't signal compatibility with the requisite
+				// protocols. e.g spv-compatibility. So if we version the api to
```

Comment:
I don't think clients need to reject orders when a server does not provide txdata. If they allow orders still, what happens is that they simply can't find the counter contract until it is mined. Even in full node, a client cannot act until the counter contract is mined. Also, in full node, if a client cannot locate the counter contract tx, it doesn't do anything but wait until the server revokes the match, iirc. Add as @chappjc noted in chat, a full node wallet can fail to locate the counter contract tx and have to wait till it is mined if a user restarts their nodes and the mempool loses the txn.

Reply:
If the above holds true, then the addition of txdata isn't really something done to aid spv-compatibility. 
As noted above whether spv or not, clients will wait for txs to be mined before acting and only abandon matches when the server revokes them, not if they cannot find a particular tx.

The real advantage to adding txdata imo, is to ensure that counter parties play their parts by broadcasting valid contracts, with receiving clients having the ability to also broadcast as a fail safe. With this addition, clients (with spv or full node wallets) can abandon matches earlier if a provided tx is invalid or fails audit checks such as address and amount. This is why I recommended in a previous review to not allude the introduction of txdata to spv compatibility - it started that way but I believe things have taken a turn since the change to allow audits run until the contract tx is mined or the server revokes the match.

Reply:
I'm not quite sure what you're getting at here. If we weren't doing SPV, we wouldn't be making these changes. 

> The real advantage to adding txdata imo, is to ensure that counter parties play their parts by broadcasting valid contracts, with receiving clients having the ability to also broadcast as a fail safe. 

We already do this verification without tx data from the server though. We only need the tx data so that SPV clients can do the same validation immediately, since they won't have access to the mempool data. The fail-safe definition strikes me wrong too. So what has failed, and why if something has failed, are we progressing the trade for the counter-party? I'm not convinced that broadcasting the counter-party's transaction is a route we want to take as a remedial action.

And even though we are no longer using broadcast timeout to secure the client, I'm not sure it's an option we should throw away. We are not very well protected against a malicious server right now. 

> With this addition, clients (with spv or full node wallets) can abandon matches earlier if a provided tx is invalid or fails audit checks such as address and amount. 

Something we can already do for full node-backed wallets. We only need the tx data so that spv clients can do it too.

I do acknowledge that with our new leniency on broadcast timeouts, having the transaction data before mining gives us little advantage. But without tx data our spv clients will be more vulnerable to attacks from malicious servers.

Which brings us back to 

> I don't think clients need to reject orders when a server does not provide txdata.

This feels like a new spin on the same conversation we've had many times. We do not want our clients sitting in the dark. A malicious server could effectively lock up the funds of every spv client with just a random hash generator.

Reply:
> We do not want our clients sitting in the dark. A malicious server could effectively lock up the funds of every spv client with just a random hash generator.

But this problem already exists, it's not particular to spv wallets. If a server sends random hashes and a client is unable to locate the tx, funds will remain locked until the server revokes the match. On that basis, sending txdata to combat this issue doesn't only apply to spv wallets.

> We already do this verification without tx data from the server though. We only need the tx data so that SPV clients can do the same validation immediately, since they won't have access to the mempool data.

Full node-backed wallets are not **always** currently able to audit contracts since they may not find the tx in their mempool for different possible reasons such as the user restarting wallet node and dexc _after_ counter party has broadcasted their tx. When this happens, the user has to wait till the contract is mined to perform any validation. Even full node-backed wallets. Sending txdata for full node-backed wallet clients gives clients the ability to perform audits on the provided txdata without needing to find the tx in the blockchain first. Agreed, I expect it'll be less common for full node-backed wallets to not find contract txs in their mempool, but it's a very realistic possibility.

> I do acknowledge that with our new leniency on broadcast timeouts, having the transaction data before mining gives us little advantage. But without tx data our spv clients will be more vulnerable to attacks from malicious servers.

The only protection full node wallets currently have over spv wallets is they can perform audit and abandon a match sooner if the contract tx doesn't meet expectations - and this advantage only exists if the full node wallet does find the contract tx before it is mined, which as has been established, may not always happen. On this note, sending txdata offers almost just as much extra advantage for full node wallet as it does for spv wallets - provide the ability to audit contracts regardless of whether they can be found in mempool or not.

Reply:
I think the meat of my point is:

> As noted above whether spv or not, clients will wait for txs to be mined before acting and only abandon matches when the server revokes them, not if they cannot find a particular tx.

This is the key reason why it makes very little difference whether a wallet can find a tx in mempool or not, whether spv or full node-backed. If we say spv wallets should not trade if they won't get txdata from the server, then we should say full node-backed wallets should abandon matches if they can't find a tx in mempool. If we allow full node wallets to wait till the tx is mined or the server revokes the match, why aren't we allowing svp wallets to do the same?

The only scenario where I think full node wallets have an advantage is the combination of the following:
- the server sends a hash of a real transaction but that does not meet audit requirements
- the wallet finds the tx (mined or unmined) and abandons the match without waiting for the server's revoke.

If that were to happen to an spv wallet that does not have the txdata, the client would not abandon the match until the server revokes it or the wallet finds the mined tx and then performs the audit checks.

Still, if a server really wants to be malicious and blind-side spv wallets, the server can provide a valid txdata that passes audit checks but is a double spend and the client will not know even if the client broadcasts the txdata because spv clients do not really have any validation that a rawtx is accepted into the blockchain until it is mined. So again, they have to wait.

Reply:
That all makes sense, and I don't disagree. Was it just the mention of spv in the TODO that you were concerned with?

Reply:
Not that we want more options and switches, but I wonder if it would be helpful to have an option for SPV-only clients to "reject orders at the funding stage when the server doesn't signal compatibility".  I think we'd have to evaluate the UX for SPV-only clients is in the absence of this data.  That is, will the confusion and suspense waiting for the initial confirm be tolerable?

Reply:
Even in the absence of the transaction data, we can still link to the transaction in a block explorer, which will presumably have it in mempool.

We could almost use dcrdata as an spv mempool, but let's not.

Reply:
@wisdom I'm just going to delete these TODOs and comments for now, if nobody objects.

Reply:
You tagged the wrong username 😉 . Life would probably be easier if my username everywhere was just [at]wisdom but we move.